src/Analysis/Engine/Impl/Infrastructure/Extensions/ArrayExtensions.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+
+namespace Microsoft.PythonTools.Analysis.Infrastructure {
+    internal static class ArrayExtensions {
```

Comment:
I think `System.Collections.Immutable` has all of this stuff. Would there be any benefit to using that (specifically `ImmutableArray`) instead of our own custom extensions to the builtin Array? I'd think it might help make clear what actually has to be immutable (and what doesn't), and it claims to optimize things to prevent unneeded copies when possible.

Reply:
Yes, `ImmutableArray` is almost free (it is a thin struct wrapper), but other types are rather slow end expensive. Referencing new package for a single type is an overkill, especially for a reference-free assembly. Also `ImmutableArray` stops being cheap when any custom change operation is required.
We can grab the sources of the `ImmutableArray`.

Reply:
I guess I'm a bit ignorant of the implications of adding a reference when we currently have none. I'm personally less satisfied by copying the source code of another library rather than using it the normal way (especially since its code seems to be non-trivial), or even leaving it how it is with some indication that uses shouldn't modify it.

Reply:
The problem may happen if several versions of assembly has to referenced through different code paths. This particular assembly (`System.Collections.Immutable`) has caused problems in RTVS because of the threading package that required special version of assembly that we couldn't use in our extension. 
So being dependent only on .net standard is better for our clients.